### PR TITLE
reader_concurrency_semaphore: Optimize resource_units destruction by postponing wait list processing

### PR DESCRIPTION
**Pull request by paszkow**
Observed 3% throughput improvement in sstable-heavy workload bounded by CPU.

SStable parsing involves lots of buffer operations which obtain and destroy resource_units. Before the patch, reosurce_unit destruction invoked maybe_admit_waiters(), which performs some computations on waiting permits. We don't really need to admit on each change of resources, since the CPU is used by other things anyway. We can batch the computation. There is already a fiber which does this for processing the _ready_list. We can reuse it for processing _wait_list as well.

The changes violates an assumption made by tests that releasing resources immediately triggers an admission check. Therefore, some of the BOOST_REQUIRE_EQUAL needs to be replaced with REQUIRE_EVENTUALLY_EQUAL as the admision check is now done in the fiber processing the _ready_list.

`perf-simple-query` --tablets --smp 1 -m 1G results obtained for fixed 400MHz frequency:
    
Before:
```
enable-cache=1
Running test with config: {partitions=10000, concurrency=100, mode=read, frontend=cql, query_single_key=no, counters=no}
Disabling auto compaction
Creating 10000 partitions...

112590.60 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   41353 insns/op,   17992 cycles/op,        0 errors)
122620.68 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   41310 insns/op,   17713 cycles/op,        0 errors)
118169.48 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   41353 insns/op,   17857 cycles/op,        0 errors)
120634.65 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   41328 insns/op,   17733 cycles/op,        0 errors)
117317.18 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   41347 insns/op,   17822 cycles/op,        0 errors)

         throughput: mean=118266.52 standard-deviation=3797.81 median=118169.48 median-absolute-deviation=2368.13 maximum=122620.68 minimum=112590.60
instructions_per_op: mean=41337.86 standard-deviation=18.73 median=41346.89 median-absolute-deviation=14.64 maximum=41352.53 minimum=41309.83
  cpu_cycles_per_op: mean=17823.50 standard-deviation=111.75 median=17821.97 median-absolute-deviation=90.45 maximum=17992.04 minimum=17713.00
```
After
```
enable-cache=1
Running test with config: {partitions=10000, concurrency=100, mode=read, frontend=cql, query_single_key=no, counters=no}
Disabling auto compaction
Creating 10000 partitions...

123689.63 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   40997 insns/op,   17384 cycles/op,        0 errors)
129643.24 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   40997 insns/op,   17325 cycles/op,        0 errors)
128907.27 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   41009 insns/op,   17325 cycles/op,        0 errors)
130342.56 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   40993 insns/op,   17286 cycles/op,        0 errors)
130294.09 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   40972 insns/op,   17336 cycles/op,        0 errors)

         throughput: mean=128575.36 standard-deviation=2792.75 median=129643.24 median-absolute-deviation=1718.73 maximum=130342.56 minimum=123689.63
instructions_per_op: mean=40993.51 standard-deviation=13.23 median=40996.73 median-absolute-deviation=3.30 maximum=41008.86 minimum=40972.48
  cpu_cycles_per_op: mean=17331.16 standard-deviation=35.02 median=17324.84 median-absolute-deviation=6.49 maximum=17383.97 minimum=17286.33
```

Improvement, not for backport.

### CODE REVIEW

**Comment by paszkow** [comment]
Replaces https://github.com/scylladb/scylladb/pull/21346

File: test/boost/reader_concurrency_semaphore_test.cc [hunk 1]
```diff
@@ -1156,14 +1156,14 @@ SEASTAR_THREAD_TEST_CASE(test_reader_concurrency_semaphore_evict_inactive_reads_
...
+    // Marking p2 as awaits should eventually allow p3 to be admitted by evicting p1
     rd2.mark_as_awaits();
-    BOOST_REQUIRE_EQUAL(semaphore.get_stats().waiters, 0);
-    BOOST_REQUIRE_EQUAL(semaphore.get_stats().need_cpu_permits, 1);
-    BOOST_REQUIRE_EQUAL(semaphore.get_stats().awaits_permits, 1);
-    BOOST_REQUIRE_EQUAL(semaphore.get_stats().inactive_reads, 0);
-    BOOST_REQUIRE_EQUAL(semaphore.get_stats().permit_based_evictions, 1);
-    BOOST_REQUIRE_EQUAL(semaphore.available_resources().count, 0);
+    REQUIRE_EVENTUALLY_EQUAL(semaphore.get_stats().waiters, 0);
+    REQUIRE_EVENTUALLY_EQUAL(semaphore.get_stats().need_cpu_permits, 1);
```
Comment (denesb) [inline test/boost/reader_concurrency_semaphore_test.cc:1162]: I think only the first needs to be `_EVENTUALLY`, once `maybe_admit_readers()` is scheduled, all the other checks should pass immediately.

Comment (paszkow) [reply test/boost/reader_concurrency_semaphore_test.cc:1162]: True, but wanted to emphasize it in case the checks get shuffled. It is also more readable, I think.

Comment (denesb) [reply test/boost/reader_concurrency_semaphore_test.cc:1162]: The checks cannot shuffle. As-is the code gives you the impression, that there can be a delay between the different checks becoming "green", whereas what really happen is that we are waiting for an event, which should make all checks pass atomically.

File: test/boost/reader_concurrency_semaphore_test.cc [hunk 2]
```diff
@@ -844,6 +844,7 @@ SEASTAR_THREAD_TEST_CASE(test_reader_concurrency_semaphore_admission) {
         auto permit2 = semaphore.obtain_permit(schema, get_name(), 1024, db::timeout_clock::now(), {}).get();
         auto irh2 = semaphore.register_inactive_read(make_empty_flat_reader_v2(s.schema(), permit2));
 
+        sleep(50ms).get();
```
Comment (avikivity) [inline test/boost/reader_concurrency_semaphore_test.cc:847]: Looks like a flaky test was born here.

File: test/boost/reader_concurrency_semaphore_test.cc [hunk 3]
```diff
@@ -863,7 +864,7 @@ SEASTAR_THREAD_TEST_CASE(test_reader_concurrency_semaphore_admission) {
         BOOST_REQUIRE_EQUAL(stats_after.reads_enqueued_for_admission, stats_before.reads_enqueued_for_admission + 1);
         BOOST_REQUIRE_EQUAL(semaphore.get_stats().waiters, 1);
 
-        std::ignore = post_enqueue_hook(cookie1);
+        [[maybe_unused]] auto value = post_enqueue_hook(cookie1);
```
Comment (avikivity) [inline test/boost/reader_concurrency_semaphore_test.cc:867]: Why this change?

File: reader_concurrency_semaphore.cc [hunk 4]
```diff
@@ -1023,7 +1025,7 @@ void reader_concurrency_semaphore::consume(reader_permit::impl& permit, resource
 
 void reader_concurrency_semaphore::signal(const resources& r) noexcept {
     _resources += r;
-    maybe_admit_waiters();
+    _ready_list_cv.signal();
```
Comment (avikivity) [inline reader_concurrency_semaphore.cc:1028]: I understand how this change improves some workloads, where a release has high chance of causing a new permit to be admitted. But can't it also regress? If a release doesn't cause a waiting request to be admitted, then we wasted a task wakeup for nothing.

Comment (avikivity) [reply reader_concurrency_semaphore.cc:1028]: I also expected to see _read_list_cv.signal() removed from maybe_admit_waiters().

Comment (avikivity) [reply reader_concurrency_semaphore.cc:1028]: Looking at maybe_admit_waiters(), I can see that moving it to execution_loop() makes sense. The two touch the same variables so one is heating the cache for the other. If they're separated by a wakeup, it goes away.

Example of a workload where this can be a regression: if the queue is usually empty. Then we waste a task to check that the queue is empty, and go back to sleep immediately.

Ideas:
1. retain maybe_admit_waiters() and make it do the minimum number of checks (perhaps none) and signal the condition variable; the old maybe_admit_waiters() can be called do_admit_waiters (and can contain the same checks, minus the conditions we moved to the new maybe_admit_waiters(), and minus the condition variable wakeup. (we have eager checks and lazy checks)
2. Consider moving the empty-queue check to the new maybe_admit_waiters()
3. Measure using perf-simple-query (it would usually have a non-empty queue so doesn't help with evaluating the empty-queue check)

My feeling is that the empty-queue check is worthwhile. Many workloads have enough resources to not queue. Checking early means we can yield to compaction or another service level or something.

Comment (denesb) [reply reader_concurrency_semaphore.cc:1028]: All single-partition reads are executed via the excution-loop, which means that all our latency sensitive reads already go through it. The loop does a poll after executing each read, so even if the poll initiated from `signal()` was deferred, at latest when the read is done, the loop polls again before going to sleep.
The exception is when the read defers, the poll might happen before the read released enough resources and the loop goes to sleep. This case wouldn't be helped with polling from `signal()`, as the read didn't release enough resources yet.
So I don't see a problem with deferring the poll.

File: test/boost/reader_concurrency_semaphore_test.cc (hunk 3 above)
Comment (paszkow) [reply test/boost/reader_concurrency_semaphore_test.cc:867]: Due to this test: https://github.com/scylladb/scylladb/blob/885e81bedc27b42138fcf838dbe234b46759198a/test/boost/reader_concurrency_semaphore_test.cc#L924-L932

and the change
https://github.com/scylladb/scylladb/pull/21918/files/885e81bedc27b42138fcf838dbe234b46759198a#diff-34cabee484b74a85b640a0fe07406cbfbce8a9722b4d3bd6bb1c639930495f36R1591

The `awaits_guard` returned from the second hook is created and destroyed if not assigned. Therefore, by the time `maybe_admit_waiters()` is called in the `execution_loop`, the permit would no longer be in the `active_await` state and as a consequence the second permit would be admitted.

The change, ensures the guard is not immediately destroyed keeping the first permit in the `active_await` state.

File: test/boost/reader_concurrency_semaphore_test.cc (hunk 2 above)
Comment (denesb) [reply test/boost/reader_concurrency_semaphore_test.cc:847]: Maybe you can add a test-only method to the semaphore, which would allow hooking into the loop wake-ups.

Comment (paszkow) [reply test/boost/reader_concurrency_semaphore_test.cc:847]: I realised, I can replace it with `REQUIRE_EVENTUALLY_EQUAL(bool(irh1), false);`

File: test/boost/reader_concurrency_semaphore_test.cc (hunk 1 above)
Comment (paszkow) [reply test/boost/reader_concurrency_semaphore_test.cc]: Fixed

File: reader_concurrency_semaphore.cc (hunk 4 above)
Comment (avikivity) [reply reader_concurrency_semaphore.cc:1028]: I'm talking about this sequence:

1. read queued
2. execution loop wakes executes read
3. read issues I/O
4. execution loop sleeps on empty queue
5. read completes, wakes execution loop
6. execution loop wakes, sees empty queue, sleeps again

This step 6 is wasteful, and not uncommon.

I'm proposing this:

1. read queued
2. execution loop wakes executes read
3. read issues I/O
4. execution loop sleeps on empty queue
5. read completes, sees queue is empty, does not wake execution loop

Basically moving the complex checks to the execution loop like the patch, but keeping the cheap check for an empty queue to avoid the wakeup.

It is very common for the queue to be empty, since the resources we reserve for reads are quite generous. It would be non-empty if we serve mostly from cache, or if serving very expensive reads like full scans.

File: test/boost/reader_concurrency_semaphore_test.cc (hunk 3 above)
Comment (avikivity) [reply test/boost/reader_concurrency_semaphore_test.cc:867]: I see. When looking not deeply enough, the two statements look similar, but std::ignore indeed causes immediate destruction.

Suggest to rename `value` to `guard` to hint to readers about what's going on.

File: reader_concurrency_semaphore.cc (hunk 4 above)
Comment (denesb) [reply reader_concurrency_semaphore.cc:1028]: Ah, I see now. This makes sense, we can add yet another maybe_something:
```
void maybe_wake_execution_loop() {
  if (!_wait_queue.empty()) {
    _ready_list_cv.notify();
}
```
Then we call this instead of notifying the condition variable directly.

Comment (denesb) [reply reader_concurrency_semaphore.cc:1028]: > Yes, we can call it maybe_admit_waiters() and hten we don't need to change the call sites.

Yes, that works too, I'm fine with that.

File: reader_concurrency_semaphore.cc [hunk 5]
```diff
@@ -1023,7 +1025,7 @@ void reader_concurrency_semaphore::consume(reader_permit::impl& permit, resource
 
 void reader_concurrency_semaphore::signal(const resources& r) noexcept {
     _resources += r;
-    maybe_admit_waiters();
+    maybe_wake_execution_loop();
```
Comment (denesb) [inline reader_concurrency_semaphore.cc:1028]: Avi suggested to repurpose `maybe_admit_waiters()` to what is `maybe_wake_execution_loop()` in this patch.
I'm think this works too.

File: test/boost/reader_concurrency_semaphore_test.cc [hunk 6]
```diff
@@ -1156,9 +1157,9 @@ SEASTAR_THREAD_TEST_CASE(test_reader_concurrency_semaphore_evict_inactive_reads_
     BOOST_REQUIRE_EQUAL(semaphore.available_resources().count, 0);
     BOOST_REQUIRE(irh1);
 
-    // Marking p2 as awaits should now allow p3 to be admitted by evicting p1
+    // Marking p2 as awaits should eventually allow p3 to be admitted by evicting p1
     rd2.mark_as_awaits();
-    BOOST_REQUIRE_EQUAL(semaphore.get_stats().waiters, 0);
+    REQUIRE_EVENTUALLY_EQUAL(semaphore.get_stats().waiters, 0);
```
Comment (denesb) [inline test/boost/reader_concurrency_semaphore_test.cc:1162]: Does this eliminate the flakyness?

**Review by denesb — APPROVED** [review]

File: reader_concurrency_semaphore.cc (hunk 5 above)
Comment (paszkow) [reply reader_concurrency_semaphore.cc:1028]: Yes indeed, but then you need to rename the current `maybe_admit_waiters` used in the `execution_loop` into something like `do_admit_waiters`.

When I was working over the code, initially it was hard to find and understand when the `execution_loop` wakes up. Thus, I thing, calling the function `maybe_wake_execution_loop` is more verbose. But I can change it, if that is unacceptable.

File: test/boost/reader_concurrency_semaphore_test.cc (hunk 6 above)
Comment (paszkow) [reply test/boost/reader_concurrency_semaphore_test.cc:1162]: This part has never caused any flakiness.

File: reader_concurrency_semaphore.cc (hunk 5 above)
Comment (denesb) [reply reader_concurrency_semaphore.cc:1028]: It is fine by me.

**Comment by denesb** [comment]
@avikivity please re-review.

**Comment by paszkow** [comment]
> ❌ - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/175/)

Reported errors:

```
Error occured during collecting of db.crt on host: artifacts-debian11-jenkins-db-node-557c9825-0-1
'NoneType' object is not iterable
Error occured during collecting of client-facing.crt on host: artifacts-debian11-jenkins-db-node-557c9825-0-1
'NoneType' object is not iterable
Error occured during collecting of cqlshrc on host: artifacts-debian11-jenkins-db-node-557c9825-0-1
'NoneType' object is not iterable
```

> #### Failed Tests (1/36625):
> * [test_mv_topology_change.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13927/testReport/junit/topology_custom/test_mv_topology_change/Tests___Unit_Tests___test_mv_topology_change_dev_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_mv_topology_change.dev.3)
> 
Existing issue: https://github.com/scylladb/scylladb/issues/21912

**Comment by mykaul** [comment]
Unit tests failed with https://github.com/scylladb/scylladb/issues/21912

Offline installer with some debian 11 issue which I have no idea what it is - and its history has vanished.

File: reader_concurrency_semaphore.cc (hunk 5 above)
Comment (avikivity) [reply reader_concurrency_semaphore.cc:1028]: Me2

File: test/boost/reader_concurrency_semaphore_test.cc (hunk 6 above)
Comment (avikivity) [reply test/boost/reader_concurrency_semaphore_test.cc:1162]: So why the change?

Comment (avikivity) [reply test/boost/reader_concurrency_semaphore_test.cc:1162]: Ah, it's documented.

**Comment by avikivity** [comment]
I asked for before/after perf-simple-query report (customary is with --smp 1 -m 1G) so we see effects on the fast-path too.

**Comment by paszkow** [comment]
> I asked for before/after perf-simple-query report (customary is with --smp 1 -m 1G) so we see effects on the fast-path too.

@avikivity  `perf-simple-query --tablets --smp 1 -m 1G` results:

Before:
```
enable-cache=1
Running test with config: {partitions=10000, concurrency=100, mode=read, frontend=cql, query_single_key=no, counters=no}
Disabling auto compaction
Creating 10000 partitions...
251338.33 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   41343 insns/op,   18467 cycles/op,        0 errors)
271951.42 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   41343 insns/op,   17507 cycles/op,        0 errors)
271525.13 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   41330 insns/op,   17522 cycles/op,        0 errors)
272440.88 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   41339 insns/op,   17477 cycles/op,        0 errors)
277071.85 tps ( 63.0 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   41341 insns/op,   17179 cycles/op,        0 errors)

throughput: mean=268865.52 standard-deviation=10048.95 median=271951.42 median-absolute-deviation=3575.35 maximum=277071.85 minimum=251338.33
instructions_per_op: mean=41339.31 standard-deviation=5.47 median=41341.14 median-absolute-deviation=3.35 maximum=41343.44 minimum=41329.91
cpu_cycles_per_op: mean=17630.26 standard-deviation=488.62 median=17506.64 median-absolute-deviation=153.16 maximum=18467.24 minimum=17178.67
```

After:
```
enable-cache=1
Running test with config: {partitions=10000, concurrency=100, mode=read, frontend=cql, query_single_key=no, counters=no}
Disabling auto compaction
Creating 10000 partitions...

255880.51 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   41030 insns/op,   18351 cycles/op,        0 errors)
253799.46 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   41024 insns/op,   18724 cycles/op,        0 errors)
249800.55 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   41024 insns/op,   18971 cycles/op,        0 errors)
263477.75 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   41013 insns/op,   18015 cycles/op,        0 errors)
264105.07 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   41023 insns/op,   18017 cycles/op,        0 errors)

throughput: mean=257412.66 standard-deviation=6223.35 median=255880.51 median-absolute-deviation=6065.08 maximum=264105.07 minimum=249800.55
instructions_per_op: mean=41022.98 standard-deviation=6.10 median=41023.74 median-absolute-deviation=1.23 maximum=41030.21 minimum=41013.28
cpu_cycles_per_op: mean=18415.82 standard-deviation=426.32 median=18351.46 median-absolute-deviation=398.80 maximum=18971.20 minimum=18015.36

```

**Comment by paszkow** [comment]
The `perf-simple-query` run with the patch, printed the following lines

```
reader_concurrency_semaphore - Semaphore user with 0/100 count and 0/21474836 memory resources: semaphore could admit new reads yet there are waiters, dumping permit diagnostics:

permits	count	memory	table/operation/state
28	0	0B	ks.cf/data-query/waiting_for_admission

28	0	0B	total

Stats:
permit_based_evictions: 0
time_based_evictions: 0
inactive_reads: 0
total_successful_reads: 3572
total_failed_reads: 0
total_reads_shed_due_to_overload: 0
total_reads_killed_due_to_kill_limit: 0
reads_admitted: 3573
reads_enqueued_for_admission: 3515
reads_enqueued_for_memory: 0
reads_admitted_immediately: 87
reads_queued_because_ready_list: 2975
reads_queued_because_need_cpu_permits: 539
reads_queued_because_memory_resources: 0
reads_queued_because_count_resources: 0
reads_queued_with_eviction: 0
total_permits: 3601
current_permits: 28
need_cpu_permits: 0
awaits_permits: 0
disk_reads: 0
sstables_read: 0
```

They originate from https://github.com/scylladb/scylladb/blob/cc3c8a3ae7cc6f0423790af8f2044c5a34c8e945/reader_concurrency_semaphore.cc#L1462-L1469

and I think we shall drop printing the diagnostics here. Admitting waiters happens now in the execution_loop and it is no longer a contradiction. So the logged information seems to be misleading. 

And I see there is one left-over. Instead of `maybe_admit_waiters();` should be `maybe_wake_execution_loop();`

**Comment by avikivity** [comment]
> > I asked for before/after perf-simple-query report (customary is with --smp 1 -m 1G) so we see effects on the fast-path too.
> 
> @avikivity `perf-simple-query --tablets --smp 1 -m 1G` results:
> 
> Before:
> 
> ```
> enable-cache=1
> Running test with config: {partitions=10000, concurrency=100, mode=read, frontend=cql, query_single_key=no, counters=no}
> Disabling auto compaction
> Creating 10000 partitions...
> 251338.33 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   41343 insns/op,   18467 cycles/op,        0 errors)
> 271951.42 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   41343 insns/op,   17507 cycles/op,        0 errors)
> 271525.13 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   41330 insns/op,   17522 cycles/op,        0 errors)
> 272440.88 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   41339 insns/op,   17477 cycles/op,        0 errors)
> 277071.85 tps ( 63.0 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   41341 insns/op,   17179 cycles/op,        0 errors)
> 
> throughput: mean=268865.52 standard-deviation=10048.95 median=271951.42 median-absolute-deviation=3575.35 maximum=277071.85 minimum=251338.33
> instructions_per_op: mean=41339.31 standard-deviation=5.47 median=41341.14 median-absolute-deviation=3.35 maximum=41343.44 minimum=41329.91
> cpu_cycles_per_op: mean=17630.26 standard-deviation=488.62 median=17506.64 median-absolute-deviation=153.16 maximum=18467.24 minimum=17178.67
> ```
> 
> After:
> 
> ```
> enable-cache=1
> Running test with config: {partitions=10000, concurrency=100, mode=read, frontend=cql, query_single_key=no, counters=no}
> Disabling auto compaction
> Creating 10000 partitions...
> 
> 255880.51 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   41030 insns/op,   18351 cycles/op,        0 errors)
> 253799.46 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   41024 insns/op,   18724 cycles/op,        0 errors)
> 249800.55 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   41024 insns/op,   18971 cycles/op,        0 errors)
> 263477.75 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   41013 insns/op,   18015 cycles/op,        0 errors)
> 264105.07 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   41023 insns/op,   18017 cycles/op,        0 errors)
> 
> throughput: mean=257412.66 standard-deviation=6223.35 median=255880.51 median-absolute-deviation=6065.08 maximum=264105.07 minimum=249800.55
> instructions_per_op: mean=41022.98 standard-deviation=6.10 median=41023.74 median-absolute-deviation=1.23 maximum=41030.21 minimum=41013.28
> cpu_cycles_per_op: mean=18415.82 standard-deviation=426.32 median=18351.46 median-absolute-deviation=398.80 maximum=18971.20 minimum=18015.36
> ```

We have a reduction in instruction count and an increase in cycles.

This can be due to poor locality, but I believe here it's noise, perhaps the machine was warmer on the second run and ran more slowly.

Let's rule out a real loss in throughput though. You can run with fixed CPU frequency (best to pin it to the lowest possible so it stays cool).

**Comment by mykaul** [comment]
Do we run perf-simple-query with PGO/LTO or without?

**Comment by paszkow** [comment]
@avikivity `perf-simple-query --tablets --smp 1 -m 1G` results obtained for fixed 400MHz frequency:

Before:
```
enable-cache=1
Running test with config: {partitions=10000, concurrency=100, mode=read, frontend=cql, query_single_key=no, counters=no}
Disabling auto compaction
Creating 10000 partitions...

112590.60 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   41353 insns/op,   17992 cycles/op,        0 errors)
122620.68 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   41310 insns/op,   17713 cycles/op,        0 errors)
118169.48 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   41353 insns/op,   17857 cycles/op,        0 errors)
120634.65 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   41328 insns/op,   17733 cycles/op,        0 errors)
117317.18 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   41347 insns/op,   17822 cycles/op,        0 errors)

         throughput: mean=118266.52 standard-deviation=3797.81 median=118169.48 median-absolute-deviation=2368.13 maximum=122620.68 minimum=112590.60
instructions_per_op: mean=41337.86 standard-deviation=18.73 median=41346.89 median-absolute-deviation=14.64 maximum=41352.53 minimum=41309.83
  cpu_cycles_per_op: mean=17823.50 standard-deviation=111.75 median=17821.97 median-absolute-deviation=90.45 maximum=17992.04 minimum=17713.00
```
After
```
enable-cache=1
Running test with config: {partitions=10000, concurrency=100, mode=read, frontend=cql, query_single_key=no, counters=no}
Disabling auto compaction
Creating 10000 partitions...

123689.63 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   40997 insns/op,   17384 cycles/op,        0 errors)
129643.24 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   40997 insns/op,   17325 cycles/op,        0 errors)
128907.27 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   41009 insns/op,   17325 cycles/op,        0 errors)
130342.56 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   40993 insns/op,   17286 cycles/op,        0 errors)
130294.09 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   40972 insns/op,   17336 cycles/op,        0 errors)

         throughput: mean=128575.36 standard-deviation=2792.75 median=129643.24 median-absolute-deviation=1718.73 maximum=130342.56 minimum=123689.63
instructions_per_op: mean=40993.51 standard-deviation=13.23 median=40996.73 median-absolute-deviation=3.30 maximum=41008.86 minimum=40972.48
  cpu_cycles_per_op: mean=17331.16 standard-deviation=35.02 median=17324.84 median-absolute-deviation=6.49 maximum=17383.97 minimum=17286.33
```

**Comment by avikivity** [comment]
Ok, with fixed frequency the results are consistent and the previous disagreement was noise.

IPC before: 2.32
IPC after: 2.36

I don't think we can read much into it. Maybe the instructions we removed had bad IPC.


Please add the performance report to the commit changelog.

**Comment by avikivity** [comment]
> Perhaps perf-simple-query should read /sys/devices/system/cpu/cpufreq/policy0/cpuinfo_min_freq or /sys/devices/system/cpu/cpufreq/policy0/base_frequency and set to it somehow (?) Does it run on cpu0 though?

It's too intrusive and requires root.

**Comment by mykaul** [comment]
> > Perhaps perf-simple-query should read /sys/devices/system/cpu/cpufreq/policy0/cpuinfo_min_freq or /sys/devices/system/cpu/cpufreq/policy0/base_frequency and set to it somehow (?) Does it run on cpu0 though?
> 
> It's too intrusive and requires root.

I don't think so. I think you can use 'scaling_setspeed' if you use the right governor. I'll look into it. I remember from https://www.alexgallego.org/perf/compiler/explorer/flatbuffers/smf/2018/06/30/effects-cpu-turbo.html that you could script it - perhaps since then there are more knobs (turbo boost? intel pstate settings?) that need to be tweaked.

**Comment by avikivity** [comment]
It requires root for sure.

**Comment by mykaul** [comment]
> It requires root for sure.

Yes. Confused root and boot...
